server/src/test/java/org/elasticsearch/env/NodeEnvironmentIT.java

```diff
@@ -123,4 +135,70 @@ public void testFailsToStartIfUpgradedTooFar() {
         assertThat(illegalStateException.getMessage(),
             allOf(startsWith("cannot upgrade a node from version ["), endsWith("] directly to version [" + Version.CURRENT + "]")));
     }
+
+    public void testUpgradeDataFolder() throws IOException, InterruptedException {
+        String node = internalCluster().startNode();
+        prepareCreate("test").get();
+        indexRandom(true, client().prepareIndex("test", "type1", "1").setSource("{}", XContentType.JSON));
+        String nodeId = client().admin().cluster().prepareState().get().getState().nodes().getMasterNodeId();
+
+        final Settings dataPathSettings = internalCluster().dataPathSettings(node);
+        internalCluster().stopRandomDataNode();
+
+        // simulate older data path layout by moving data under "nodes/0" folder
+        final List<Path> dataPaths = Environment.PATH_DATA_SETTING.get(dataPathSettings)
+            .stream().map(PathUtils::get).collect(Collectors.toList());
+        dataPaths.forEach(path -> {
+                final Path targetPath = path.resolve("nodes").resolve("0");
+                try {
+                    Files.createDirectories(targetPath);
+
+                    try (DirectoryStream<Path> stream = Files.newDirectoryStream(path)) {
+                        for (Path subPath : stream) {
+                            String fileName = subPath.getFileName().toString();
+                            Path targetSubPath = targetPath.resolve(fileName);
+                            if (fileName.equals("nodes") == false) {
+                                Files.move(subPath, targetSubPath, StandardCopyOption.ATOMIC_MOVE);
+                            }
+                        }
+                    }
+                    IOUtils.fsync(targetPath, true);
+                    IOUtils.fsync(path, true);
```

Comment:
I don't think we care about `fsync`s here.

Reply:
fixed in b139f14